### PR TITLE
Generate L1Blocks system tx

### PR DESCRIPTION
**Pull request by icemelon**
## 1. Purpose or design rationale of this PR

Generate L1Blocks system tx

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [x] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: core/types/system_tx.go [hunk 1]
```diff
@@ -10,7 +10,8 @@ import (
 
 type SystemTx struct {
 	Sender common.Address  // pre-determined sender
-	To     *common.Address // system contract
+	To     common.Address  // system contract address
+	Nonce  uint64          // nonce of sender account
```
Comment (Thegaram) [inline core/types/system_tx.go]: Why add nonce? Can't we just bump it implicly as we do for `L1MessageTx`?

File: core/types/system_tx.go [hunk 2]
```diff
@@ -22,7 +23,7 @@ func (tx *SystemTx) txType() byte { return SystemTxType }
 func (tx *SystemTx) copy() TxData {
 	return &SystemTx{
 		Sender: tx.Sender,
-		To:     copyAddressPtr(tx.To),
```
Comment (Thegaram) [inline core/types/system_tx.go:25]: The doc above says that "copy creates a deep copy". (Though this is also not respected for `BlobTx`.)

File: rollup/system_contracts/l1_blocks.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+	return &types.SystemTx{
+		Sender: rcfg.SystemSenderAddress,
+		To:     rcfg.L1BlocksAddress,
+		Nonce:  nonce,
+		Data:   data,
+	}, nil
+}
+
+func (w *L1BlocksWorker) GenerateL1BlockMsgs(from, to uint64, state StateDB) ([]*types.SystemTx, error) {
```
Comment (Thegaram) [inline rollup/system_contracts/l1_blocks.go]: This method will introduce a lot of latency into block building, probably not suitable for production use. I suggest that `L1BlocksWorker` fetch blocks proactively (stored in memory) and returns finalized blocks from this in-memory buffer. We can change this later.

File: core/types/system_tx.go (hunk 1 above)
Comment (icemelon) [reply core/types/system_tx.go]: to make it easier to track the tx ordering. `L1MessageTx` has the L1 msg index to track the ordering but `SystemTx` has none if there's no nonce

Comment (Thegaram) [reply core/types/system_tx.go]: But the sender (sequencer) has full control over the ordering anyway. And we can add additional checks in the contract itself.

Comment (icemelon) [reply core/types/system_tx.go]: removed

File: rollup/system_contracts/l1_blocks.go (hunk 3 above)
Comment (icemelon) [reply rollup/system_contracts/l1_blocks.go]: moved it to sync service now

**Review by vyzo — COMMENTED** [review]
looks reasonable at first pass.

Let me think a bit and do another pass later tonight, but i dont see anything wrong with it.

**Comment by icemelon** [comment]
I'll merge this PR first to test it out. We can make changes or fixes in the subsequent prs.

File: core/state_transition.go [hunk 4]
```diff
@@ -273,6 +273,18 @@ func (st *StateTransition) buyGas() error {
 
 func (st *StateTransition) preCheck() error {
 	if st.msg.IsSystemTx() {
+		// Make sure this transaction's nonce is correct.
```
Comment (Thegaram) [inline core/state_transition.go:276]: These checks are not needed since the nonce field was removed.

File: rollup/sync_service/sync_service.go [hunk 5]
```diff
@@ -145,13 +175,95 @@ func (s *SyncService) SubscribeNewL1MsgsEvent(ch chan<- core.NewL1MsgsEvent) eve
 	return s.scope.Track(s.msgCountFeed.Subscribe(ch))
 }
 
+func (s *SyncService) SubscribeNewL1BlocksTx(ch chan<- core.NewL1MsgsEvent) event.Subscription {
+	return s.scope.Track(s.l1BlocksPool.l1BlocksFeed.Subscribe(ch))
+}
+
+func (s *SyncService) CollectL1BlocksTxs(latestL1BlockNumberOnL2, maxNumTxs uint64) []*types.SystemTx {
```
Comment (Thegaram) [inline rollup/sync_service/sync_service.go:182]: Will this method be able to handle L2 reorgs? E.g. when we've marked some L1 blocks as included (removed from `pendingL1BlocksTxs`) but then some blocks are reorged.

**Pull request merged**